src/Store.ts

```diff
@@ -0,0 +1,392 @@
+		const Model = this.Model;
+		if (Model && object) {
+			const prototype = Model.prototype;
+			const restore = prototype._restore;
+			if (restore) {
+				// the prototype provides its own restore method
+				object = restore.call(object, Model, mutateAllowed);
+			} else if (hasProto && mutateAllowed) {
+				// the fast easy way
+				// http://jsperf.com/setting-the-prototype
+				object['__proto__'] = prototype;
+			} else {
+				// create a new object with the correct prototype
+				object = <Hash<any>> lang.create(prototype, object);
+			}
+		}
+		return object;
+	}
+
+	/**
+	 * Sets an object's identity description:
+	 * This method sets an object's identity and is useful to override to support
+	 * multi-key identities and object's whose properties are not stored directly on the object.
+	 *
+	 * @param object The target object
+	 * @param identityArg The argument used to set the identity
+	 */
+	protected _setIdentity(object: { [ index: string ]: any; set: (...args: any[]) => any}, identityArg: string | number) {
+		if (object.set) {
+			object.set(this.idProperty, identityArg);
+		} else {
+			object[this.idProperty] = identityArg;
+		}
+	}
+
+	/**
+	 * This creates a new instance from the store's model.
+	 * @param properties The properties that are passed to the model constructor to
+	 * be copied onto the new instance. Note, that should only be called
+	 * when new objects are being created, not when existing objects
+	 * are being restored from storage.
+	 */
+	create(properties: {}) {
+		return new this.Model(properties);
+	}
+
+	/**
+	 * Retrieves a range of objects from the collection, returning a promise to an array.
+	 *
+	 * @param args Contains the starting index of objects to return (0-indexed)
+	 * and the exclusive end of objects to return
+	 * @return A FetchPromise that will resolve with the results of the fetch
+	 */
+	fetchRange(args: dstore.FetchRangeArgs):dstore.FetchPromise<T> {
+		throw new Error("This method is abstract");
+	}
+
+	filter<T>(...args: any[]): dstore.Collection<T> {
+		return QueryMethod(<QueryMethodArgs<T>> {
+			type: 'filter',
+			normalizeArguments: function (filter) {
+				const Filter = this.Filter;
+				if (filter instanceof Filter) {
+					return [ filter ];
+				}
+				return [ new Filter(filter) ];
+			}
+		}).apply(this, args);
+	}
+
+	/**
+	 * Iterates over the query results, based on
+	 * https://developer.mozilla.org/en/Core_JavaScript_1.5_Reference/Objects/Array/forEach.
+	 * Note that this executes asynchronously, and returns a promise to indicate when the
+	 * callback has been executed.
+	 *
+	 * @param callback Function that is called for each object in the query results
+	 * @parma thisObject The object to use as |this| in the callback.
+	 * @returns A Promise indicating when the operatio is done
+	 */
+	forEach<T>(callback: (item: T, index: number | string, collection: T[]) => void, thisObject?: any) {
+		return this.fetch().then((data) => {
+			if (data) {
```

Comment:
It wasn't necessary, I thought it would be safer this way since we're trying to access `data[0]` below without making sure `data` exists, but since the return type of fetch is now statically guaranteed it can just be removed.